docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,240 @@
+variables and other information at runtime to seed Conduit's services with 
+the necessary information for full operation. 
+
+## Goals 
+
+- Fully load and configure pipelines from a file.
+- Properly handle environment variable injection into that file.
+
+### Non Goals
+
+Defining what we don't want out of this design and implementation is equally 
+important as what we _do_ want.
+
+- We don't care about the structure or schema of configuration files.
+- We shouldn't implement key management or encryption for secrets.
+
+## Scope
+
+We are strictly concerned with the loading of pipeline files at start time and 
+how those files are handled through Conduit's full operation cycle.
+
+Configuration file schema and other file-level design choices are out of scope.
+
+This document doesn't make any definitions or statements about how the UI should
+handle these changes. Those should be reserved for later design documents and
+discussion.
+
+## Key Terms 
+
+**pipeline** - a Conduit pipeline with `n` number of Source connectors, 
+`m` number of Processors, and `o` number of Destination connectors.
+
+## Context 
+
+Running Conduit in production environments requires a strict set of
+guarantees. Manually configuring a Conduit server in a GUI and keeping those
+high standards places more work on the end user, reduces confidence in the
+pipeline's integrity, and in the worst case could result in corrupted or lost
+data.
+
+This document expands on previous discussions to create a production mode of
+Conduit that either doesn't serve the application GUI or serves a functionally 
+limited version of the GUI instead.
+
+Conduit Issues #32 and #33 describe two different features to address some of 
+these concerns, but there are additional design decisions to make.
+
+## Design 
+
+We propose the creation of a single configuration file that contains all of
+Conduit's configuration values as well as those for `n` number of Pipelines.
+We consider a pipeline as `m` number of Connectors and `o` number of Processors.
+
+We propose naming this file `conduit.*` and it's default location should be
+at the root of the project. The format of the file is to be determined, though
+`yaml` has been proposed previously.
+
+If this file is passed at start time, we propose that Conduit run in production 
+mode. Production mode initially will not serve the UI. In the future, it would 
+serve a read-only optimized view of Conduit.
+
+Configuration must securely accept secret information. We don't want to solve
+the problem of secrets management, but we must be in the business of secret
+concealment. Additionally, we already allow environment variables so we must 
+continue handling those correctly. 
+
+We propose merging environment variables with the user's provided configuration 
+file at runtime and replacing templated variables in the configuration file with 
+their secret values at runtime.
```

Comment:
How about CLI args? For example, what if we have the HTTP port specified in the config file and through the CLI arg too? 

And a nitpick: it might be good to be explicit about the config priorities (config file > environment variable).

Reply:
Good point! I updated the document to reflect this. I think we should merge environment variables and CLI flags down into a cohesive `Environment` object that can contain sensitive information.

Reply:
IMO the priority should be CLI flags (highest priority) > environment variables (lower priority) > configuration file (lowest priority).

That said, I feel like we are mixing apples and oranges. CLI flags would only apply to the main Conduit configuration file (which we don't yet support, relevant issue: https://github.com/ConduitIO/conduit/issues/31), while this document is describing the pipeline configuration file. I don't think CLI flags would apply to pipeline configuration files, they should be able to reference environment variables though.

Reply:
I agree about both priority and that the CLI flags is an overlapping but unrelated item; I only want this to concern itself with loading environment variables necessary for pipeline files, not how they're managed or set. I can update this to be more specific about scope around environment variable handling and set a non-goal for CLI flag settings or anything like that.